File: 02/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<16)
	writer := bufio.NewWriterSize(os.Stdout, 1<<16)
	defer writer.Flush()

	// Читаем 10 чисел
	nums := make([]int, 10)
	for i := 0; i < 10; i++ {
		line, _ := reader.ReadString('\n')
		nums[i], _ = strconv.Atoi(strings.TrimSpace(line))
	}

	result := solve(nums)
	writer.WriteString(fmt.Sprintf("%d\n", result))
}

// solve находит сумму подмножества, ближайшую к 100
// При равном расстоянии выбирает большую сумму
func solve(nums []int) int {
	const target = 100
	bestSum := 0
	bestDist := target // расстояние от 0 до 100

	// Перебираем все 2^10 = 1024 подмножества
	for mask := 0; mask < (1 << 10); mask++ {
		sum := 0
		for i := 0; i < 10; i++ {
			if mask&(1<<i) != 0 {
				sum += nums[i]
			}
		}

		dist := abs(sum - target)

		// Выбираем лучшую сумму:
		// - меньшее расстояние до 100
		// - при равном расстоянии — большую сумму
		if dist < bestDist || (dist == bestDist && sum > bestSum) {
			bestSum = sum
			bestDist = dist
		}
	}

	return bestSum
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: 02/main.rs
use std::io::{self, BufRead, BufWriter, Write};

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Читаем 10 чисел
    let mut nums = Vec::with_capacity(10);
    for _i in 0..10 {
        let mut line = String::new();
        reader.read_line(&mut line).unwrap();
        let num: i32 = line.trim().parse().unwrap();
        nums.push(num);
    }

    let result = solve(&nums);
    writeln!(writer, "{}", result).unwrap();
    writer.flush().unwrap();
}

// solve находит сумму подмножества, ближайшую к 100
// При равном расстоянии выбирает большую сумму
fn solve(nums: &[i32]) -> i32 {
    const TARGET: i32 = 100;
    let mut best_sum = 0;
    let mut best_dist = TARGET; // расстояние от 0 до 100

    // Перебираем все 2^10 = 1024 подмножества
    for mask in 0..(1 << 10) {
        let mut sum = 0;
        for i in 0..10 {
            if mask & (1 << i) != 0 {
                sum += nums[i];
            }
        }

        let dist = (sum - TARGET).abs();

        // Выбираем лучшую сумму:
        // - меньшее расстояние до 100
        // - при равном расстоянии — большую сумму
        if dist < best_dist || (dist == best_dist && sum > best_sum) {
            best_sum = sum;
            best_dist = dist;
        }
    }

    best_sum
}





